File: app/src/main/java/com/example/mobile_challenge/main/MainViewModel.kt
package com.example.mobile_challenge.main

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mobile_challenge.App
import com.example.mobile_challenge.R
import com.example.mobile_challenge.db.AppDatabase
import com.example.mobile_challenge.model.*
import com.example.mobile_challenge.utility.ClientApi
import com.example.mobile_challenge.utility.SingleLiveData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.UnstableDefault
import java.util.*

@OptIn(UnstableDefault::class)
class MainViewModel(
  private var client: ClientApi = App.clientApi,
  private val dataBase: AppDatabase = App.db,
  private var quotes: List<QuoteEntity> = arrayListOf(),
  var currencyList: ArrayList<CurrencyEntity> = arrayListOf()
) : ViewModel() {

  // Handle Error
  val error: SingleLiveData<Int> = SingleLiveData()

  // Live currency value
  private val _liveValue: MutableLiveData<Double> = MutableLiveData()
  val liveValue: LiveData<Double> = _liveValue

  // Update "FROM" button
  private val _fromCode: MutableLiveData<String> = MutableLiveData()
  val fromCode: LiveData<String> = _fromCode

  // Update "TO" button
  private val _toCode: MutableLiveData<String> = MutableLiveData()
  val toCode: LiveData<String> = _toCode

  private var currencyFrom = 1.0
  private var currencyTo = 1.0

  init {
    getCurrency()
    getQuote()
  }

  // Handle Currency

  private fun getCurrency() {
    viewModelScope.launch(Dispatchers.Default) {
      when (val response = client.httpRequestGetList()) {
        is ResponseOptionsCurrency.SuccessResponse -> {
          handleCurrencySuccess(response)
        }
        is ResponseOptionsCurrency.ErrorResponse -> {
          error.postValue(R.string.connection_error)
          setCurrencyFromDb()
        }
      }
    }
  }

  private fun handleCurrencySuccess(response: ResponseOptionsCurrency.SuccessResponse) {
    if (response.data.success && response.data.currencies.isNotEmpty()) {
      updateCurrencyTableDb(response.data)
    } else {
      error.postValue(R.string.fetch_data_error)
      setQuoteFromDb()
    }
  }

  private fun updateCurrencyTableDb(currencyList: CurrencyResponse) {
    viewModelScope.launch(Dispatchers.Default) {
      val currenciesMap = currencyList.currencies
      val list = arrayListOf<CurrencyEntity>()
      var id = 1
      currenciesMap.forEach { entry ->
        val currency = CurrencyEntity(
          id,
          entry.key,
          entry.value
        )
        list.add(currency)
        id++
        dataBase.currencyDao().upsert(currency)
      }
      this@MainViewModel.currencyList = list
    }
  }

  private fun setCurrencyFromDb() {
    viewModelScope.launch(Dispatchers.Default) {
      val list = dataBase.currencyDao().getAll()
      if (list.isNotEmpty()) {
        list.flatMapTo(currencyList) { arrayListOf(it) }
      }
    }
  }

  // Handle Quote

  private fun getQuote() {
    viewModelScope.launch(Dispatchers.Default) {
      when (val response = client.httpRequestGetLive()) {
        is ResponseOptionsQuote.SuccessResponse -> {
          handleQuoteSuccess(response)
        }
        is ResponseOptionsQuote.ErrorResponse -> {
          error.postValue(R.string.connection_error)
          setQuoteFromDb()
        }
      }
    }
  }

  private fun handleQuoteSuccess(response: ResponseOptionsQuote.SuccessResponse) {
    if (response.data.success && response.data.quotes.isNotEmpty()) {
      updateQuoteTableDb(response.data)
      setQuoteValue("BRL", "TO")
    } else {
      error.postValue(R.string.fetch_data_error)
      setQuoteFromDb()
    }
  }

  private fun updateQuoteTableDb(data: QuoteResponse) {
    val list = mutableListOf<QuoteEntity>()
    var id = 1
    data.quotes.forEach { entry ->
      val quote = QuoteEntity(
        id,
        entry.key.substring(0..2),
        entry.key.substring(3..5),
        entry.value
      )
      list.add(quote)
      id++
      dataBase.quoteDao().upsert(quote)
    }
    quotes = list
  }

  private fun setQuoteFromDb() {
    viewModelScope.launch(Dispatchers.Default) {
      val quoteList = dataBase.quoteDao().getAll()
      if (quoteList.isNotEmpty()) {
        quotes = quoteList
        setQuoteValue("BRL", "TO")
      }
    }
  }

  fun setQuoteValue(code: String, type: String) {
    val value = quotes.find { it.to == code }?.value
    value?.let {
      if (type == "FROM") {
        currencyFrom = 1 / value
        _fromCode.postValue(code)
      } else {
        currencyTo = value
        _toCode.postValue(code)
      }
      _liveValue.postValue(currencyFrom * currencyTo)
    }
  }

}

File: mobile-challenge/Main/MainViewModel.swift
//
//  MainViewModel.swift
//  mobile-challenge
//
//  Created by Kivia on 5/2/20.
//  Copyright © 2020 AP Club. All rights reserved.
//

import Foundation

class MainViewModel {
  
  // Handle Error
  let error = Observable<String>()
  
  // Live currency value
  let liveValue = Observable<Double>()
  
  // Update "FROM" button
  let fromCode = Observable<String>()
  
  // Update "TO" button
  let toCode = Observable<String>()
  
  var currencyList = [] as [CurrencyEntity]
  var quotes = [] as [QuoteEntity]
  var currencyFrom = 1.0
  var currencyTo = 1.0
  var dataBase : AppDataBase?
  
  init(application: AppDelegate) {
    dataBase = application.dataBase
    guard let clientApi = application.clientApi else { return }
    getQuote(clientApi)
    getCurrency(clientApi)
  }
  
  private func getCurrency(_ clientApi: ClientApi) {
    clientApi.fetchCurrencyList { (response) in
      switch (response) {
      case .SucessResponse(let data): do {
        self.handleCurrencySucess(data)
        }
      case .ErrorResponse(let message): do {
        self.error.property = message
        self.setCurrencyFromDb()
        print(message)
        }
      }
    }
  }
  
  private func handleCurrencySucess(_ data: (CurrencyResponse)) {
    if data.success || !data.currencies.isEmpty{
      self.updateCurrencyTableDB(data: data)
    } else {
      self.error.property = "Error while fetching data"
      self.setCurrencyFromDb()
      print("False Sucess")
    }
  }
  
  private func updateCurrencyTableDB(data: CurrencyResponse){
    let sortedCurrency = data.currencies.sorted(by: { (first, second) -> Bool in
      first.key < second.key
    })
    var id = 1
    var list = [] as [CurrencyEntity]
    sortedCurrency.forEach { (dic) in
      let currency = CurrencyEntity(
        id,
        dic.key,
        dic.value)
      list.append(currency)
      id+=1
    }
    self.currencyList = list
    // Update Data base
    self.dataBase?.insertOrUpdateListCurrencyEntityTable(list: list)
  }
  
  private func setCurrencyFromDb() {
    print("DATABASE \(String(describing: dataBase))")
    let list = dataBase?.selectAllCurrencyEntityTable()
    print("CURRENCY LIST \(String(describing: list))")
    if (list!.count > 0) {
      currencyList = list!
    }
  }
  
  private func getQuote(_ clientApi: ClientApi) {
    clientApi.fetchQuoteLive { (response) in
      switch (response) {
      case .SucessResponse(let data): do {
        self.handleQuoteSucess(data)
        }
      case .ErrorResponse(let message): do {
        self.error.property = message
        self.setQuoteFromDb()
        print(message)
        }
      }
    }
  }
  
  private func handleQuoteSucess(_ data: (QuoteResponse)) {
    if data.success || !data.quotes.isEmpty{
      updateQuoteTableDB(data: data)
      setQuoteValue(code: "BRL", type: "TO")
    } else {
      self.error.property = "Error while fetching data"
      self.setQuoteFromDb()
      print("False Sucess")
    }
  }
  
  func updateQuoteTableDB(data: QuoteResponse) {
    let sortedQuotes = data.quotes.sorted(by: { (first, second) -> Bool in
      first.key < second.key
    })
    var id = 1
    var quoteList = [] as [QuoteEntity]
    sortedQuotes.forEach { (quote) in
      let quoteEntity = QuoteEntity(
        id,
        quote.key.substring(to: 3),
        quote.key.substring(from: 3),
        quote.value)
      quoteList.append(quoteEntity)
      id+=1
    }
    self.quotes = quoteList
    // update database
    self.dataBase?.insertOrUpdateListQuoteEntityTable(list: quoteList)
  }
  
  private func setQuoteFromDb() {
    let quoteList = dataBase?.selectAllQuoteEntityTable()
    print("QUOTE LIST \(String(describing: quoteList))")
    if (quoteList!.count > 0) {
      self.quotes = quoteList!
      setQuoteValue(code: "BRL", type: "TO")
    }
  }
  
  func setQuoteValue(code: String, type: String) {
    let value = self.quotes.filter { (quote) -> Bool in quote.to == code }
    print("QUOTE value \(String(describing: value))")
    if (value as [QuoteEntity]?) != nil && value.count > 0 {
      if (type == "FROM") {
        self.currencyFrom = 1 / value[0].value
        fromCode.property = (code)
      } else {
        self.currencyTo = value[0].value
        toCode.property = (code)
      }
      liveValue.property = (currencyFrom * currencyTo)
    }
  }
}
